src/core_plugins/kibana/public/dashboard/actions/embeddables.js

```diff
@@ -1,58 +1,74 @@
-  /**
-   * @param panelId {string}
-   * @param embeddable {Embeddable}
-   */
-  (panelId, embeddable) => ({ embeddable, panelId })
-);
-
-export const embeddableRenderError = createAction('EMBEDDABLE_RENDER_ERROR',
-  /**
-   *
-   * @param panelId {string}
-   * @param error {string|object}
-   */
-  (panelId, error) => ({ panelId, error })
-);
+import {
+  updatePanel
+} from './panels';
+
+import {
+  getPanel,
+  getEmbeddablePersonalization,
+} from '../../selectors/dashboard_selectors';
+
+export const embeddableIsInitializing = createAction('EMBEDDABLE_IS_INITIALIZING');
+export const embeddableInitialized = createAction('EMBEDDABLE_INITIALIZED');
+export const setStagedFilter = createAction('SET_STAGED_FILTER');
+export const clearStagedFilters = createAction('CLEAR_STAGED_FILTERS');
+export const embeddableError = createAction('EMBEDDABLE_ERROR');
 
 /**
+ * The main point of communication from the embeddable to the dashboard. Any time state in the embeddable
+ * changes, this function will be called. The data is then extracted from EmbeddableState and stored in
+ * redux so the appropriate actions are taken and UI updated.
  *
+ * @param {function} dispatch - redux's dispatch function
+ * @param {function} getState - redux's getState function
+ * @param {string} panelId - the id of the panel whose state has changed.
+ * @param {EmbeddableState} embeddableState - the new state of the embeddable.
+ */
+function embeddableStateChanged(dispatch, getState, panelId, embeddableState) {
+  // Translate embeddableState to things redux cares about.
+  const personalization = getEmbeddablePersonalization(getState(), panelId);
+  if (!_.isEqual(embeddableState.personalization, personalization)) {
+    const panel = getPanel(getState(), panelId);
+    dispatch(updatePanel({ ...panel, embeddableConfig: _.cloneDeep(embeddableState.personalization) }));
+  }
+
+  if (embeddableState.stagedFilter) {
+    dispatch(setStagedFilter({ stagedFilter: embeddableState.stagedFilter, panelId }));
+  }
+
+  // TODO: communicate time range and filter changes from the embeddable through to dashboard rather than
+  // through angular components.
+  // const timeRange = getTimeRange(getState());
+  // if (!_.isEqual(timeRange, embeddableState.timeRange)) {
+  //   dispatch(updateTimeRange({ ...panel, embeddableConfig: embeddableState.personalization }));
+  // }
+}
+
+/**
+ * Creates the embeddable and registers it in the embeddableHandlerCache.
  * @param embeddableFactory {EmbeddableFactory}
- * @param panelElement {Node}
- * @param panel {PanelState}
- * @param containerApi {ContainerAPI}
+ * @param panelId {string}
  * @return {function(*, *)}
  */
-export function renderEmbeddable(embeddableFactory, panelElement, panel, containerApi) {
-  return (dispatch) => {
+export function initializeEmbeddable({ embeddableFactory, panelId }) {
+  return (dispatch, getState) => {
+    const panel = getPanel(getState(), panelId);
```

Comment:
So, I think the one downside to your proposed approach is that we'd be moving the async call into the react component, which I've been trying to get away from because some deem it an anti pattern.  If you look at this: https://codebrahma.com/structuring-async-operations-react-redux-applications/ , this persons `How we can improve from here ?` part where they say to use redux and middleware for the async calls seems to be the pattern I'm following here.

If it weren't for the functions stored in the cache, I could put the rest of the metadata in redux and don't have a problem with that.

Is there an approach I'm missing where I can keep the async fetches in the actions, but still get rid of the cache here?

Reply:
> n DashboardPanel's componentDidMount (maybe), we'd create the underling embeddable and insert it into the DOM, passing it the initialization data that was retrieved by connect's mapStateToProps, and passing it a state change handler that would look something like (state) => this.onEmbeddableStateChange(state) or something

It's this part that I think I disagree with.  I like that initialize returns an Embeddable instance with the render, destroy and onPanelStateChanged functions.  I like that I can call onPanelStateChanged without needing the embeddable to even be rendered (so it's kept up to date for when it eventually is rendered on screen).  I also like that I don't have to pass the embeddableFactory along.  I think what you are proposing is having `embeddableFactory.initialize(id, data) => metadata` then later calling `embeddableFactory.render(id, data, onContainerStateChanged) => onEmbeddableStateChanged`  Having two calls on the embeddableFactory feels odd to me.  I like the single create call which returns an Embeddable instance for all our needs.

I could remove the render method on the embeddableHandlerCache and just let anyone grab the Embeddable object, so that all actual render and destroy calls are inside the react component, though anyone could still grab the embeddable object and call render/destroy wherever they wanted.  But maybe that wouldn't feel as wrong?  So instead of
```
  async componentDidMount() {
    embeddableHandlerCache.render(this.props.panelId, this.panelElement);
  }
```

it'd be
```
  async componentDidMount() {
    embeddableHandlerCache.get(this.props.panelId).render(this.panelElement);
  }
```

Essentially the same thing but at least all references to "render" (and we could do the same with "destroy" are in the component?